src/common/static_ptr.h

```diff
@@ -33,74 +33,91 @@ namespace _mem {
 enum class op {
   copy, move, destroy, size
 };
-template<typename T>
+// B is the base pointer type of the class that first created the
+// function/datum. It's only used to control copy/move support and
+// noexcept annotation. Since the ward in static_ptr_base only allows
+// moves/copies from more permissive to less, we should never end up
+// with a situation where static_ptr thinks we should be able to copy
+// or move something that can't unless people abuse
+// reinterpret_pointer_cast. But it's a reinterpreting cast. They know
+// what they're getting into.
+template<typename B, typename T>
 static std::size_t op_fun(op oper, void* p1, void* p2)
-{
+  noexcept((!std::is_copy_constructible_v<B> ||
+	    std::is_nothrow_copy_constructible_v<B>) &&
+	   (!std::is_move_constructible_v<B> ||
+	    std::is_nothrow_move_constructible_v<B>)) {
   auto me = static_cast<T*>(p1);
 
-  switch (oper) {
-  case op::copy:
-    // One conspicuous downside is that immovable/uncopyable functions
-    // kill compilation right here, even if nobody ever calls the move
-    // or copy methods. Working around this is a pain, since we'd need
-    // four operator functions and a top-level class to
-    // provide/withhold copy/move operations as appropriate.
-    new (p2) T(*me);
-    break;
-
-  case op::move:
-    new (p2) T(std::move(*me));
-    break;
-
-  case op::destroy:
+  if constexpr (std::is_copy_constructible_v<B>) {
+    if (oper == op::copy) {
+      new (p2) T(*me);
+      return 0;
+    }
+  }
+
+  if constexpr (std::is_move_constructible_v<B>) {
+    if (oper == op::move) {
+      new (p2) T(std::move(*me));
+      return 0;
+    }
+  }
+
+  if (oper == op::destroy) {
     me->~T();
-    break;
+    return 0;
+  }
 
-  case op::size:
+  if (oper == op::size) {
     return sizeof(T);
   }
-  return 0;
-}
+  std::terminate();
```

Comment:
Mostly because when I throw exceptions, even if only so I can check for them in test code, someone tells me to take it back out :)

Reply:
Maybe they'll have a change of heart. ;-) It seems like the logical thing to do here, at least giving calling code a chance to capture it and let the stack unwind, rather than stopping the program because this failed. (Even if it eventually stops the program anyway.)